### DISCUSSION TITLE
Many issues arise if mysql uses option MYSQLI_OPT_INT_AND_FLOAT_NATIVE

### DISCUSSION THREAD

**Discussion by mc12345678** [zencart/zencart#5821] [Ideas]
While I haven't found anywhere in the issues/discussions about this, a huge undertaking has been put forwards to try to force/ensure type comparison with all data operations. While I may not be using the standard verbiage, I am in the midst of stepping through a ZC 1.5.8. installed store with demo product to see what issues I may be able to raise.

In driving for further development using PHP 8.2, I had done some research on how/why database stored `integers` were returned as `string` text. Further, it seemed that there were ways to alter that so that the data provided back from the database query may be more "coding" friendly. I stumbled across the database option: `MYSQLI_OPT_INT_AND_FLOAT_NATIVE` which can be set on a per load type basis such that `int` and `float` data types are provided to php code as well `int` and `float`, applicably.

Well, I ran my initial testing a few weeks ago, to include setting that flag for all of Zen Cart operations. Today, I go to admin->catalog->categories/products and the first thing I notice is that all of the categories and all of the product are marked as disabled...

Basically because of:
https://github.com/zencart/zencart/blob/d4398a6ccc9e1c8f30007a0720d9e6024c123109/admin/category_product_listing.php#L951

If I change the comparison like so:
````diff
-                  if ($product['products_status'] === '1') {
+                  if ($product['products_status'] == '1') {
````
Then "operation" is restored...

Now, this option is basically only available on mysqlnd, which does not appear to be a standard/required set of code for php (though installed in a default Windows box).

So, perhaps this is to asking the same question again, how far is this strict comparison coding going to go and what is being done to prevent installation while `MYSQLI_OPT_INT_AND_FLOAT_NATIVE` is implemented? I'm planning on disabling this useful feature considering how much has been done to force almost everything to be string based (without also forcing these potential numbers to be type-cast to strings).

**Comment by scottcwilson** [comment]
Sounds like we had better check it and ensure it is disabled.

**Comment by proseLA** [comment]
i would love to see the db class return boolean, float, integer, etc as opposed to strictly strings.  i agree that would be more coder friendly.

as php develops, it seems strict typing will be the norm.  and i am NOT a fan of moving the code base backwards.

in the example above, i believe said code would need to get modified as so:

```php
if ($product['products_status'] === 1) {
```
lets try and embrace the future as opposed to hanging onto the past.

**Comment by torvista** [comment]
The idea of inserting an integer, into a db field defined as Integer, but it back comes out as a string is an aberration/looks like a bug. That's like putting in utf8 and getting out iso8859.   

Common sense says if it's possible to correct this historical error it should be done.
It would be progress towards bringing the haphazard type casting everywhere, under control.

Currently, modified files are getting strict typing/comparisons as in the above example but based on whatever the variable happens to be at that point, not what it should be.
Continuing to do that without implementing MYSQLI_OPT_INT_AND_FLOAT_NATIVE already, means the code becomes more fragile and will break more if/when MYSQLI_OPT_INT_AND_FLOAT_NATIVE is ever adopted.

Sooner the better in my opinion, or it's just kicking the can down the road.

**Comment by lat9** [comment]
> Continuing to do that without implementing MYSQLI_OPT_INT_AND_FLOAT_NATIVE already, means the code becomes more fragile and will break more if/when MYSQLI_OPT_INT_AND_FLOAT_NATIVE is ever adopted.

I'm not sure what's meant by _is ever adopted_, but that feature has been available int the `mysqli_` support since at least PHP 5.6.

I _thought_ that one of the drivers for going to using `===` vs. `==` was for improved performance, but I'm finding conflicting reports when searching the web.  If anyone can find a conclusive posting that either validates or invalidates that performance improvement, I'd be much appreciative.

If it turns out that the performance-improvement is negligible, we could have the `$db` object's instantiation set `MYSQLI_OPT_INT_AND_FLOAT_NATIVE` to 1, but I'll note that that would be an extremely 'breaking change' for not only the current Zen Cart core, but also a plethora of plugins that have adopted the use of `===` for database query outputs.

Unfortunately, with [this](https://github.com/zencart/zencart/commit/2d3eb5756de29e14d1331e8fc6c53defb3aaaf62) change (which, FWIW, I don't agree with), it's possible that non-core code could use the [`mysqli_options`](https://www.php.net/manual/en/mysqli.options.php) function to revert the database query output back to the current character representation.    I'll note that the `mysqli_` function-set provides no means to 'get' the options currently being used, nor does that updated option get reflected in a `SELECT @@sql_mode` query.
